README.md

```diff
@@ -198,289 +189,492 @@ e.increment();  // logs 1
+        for (const [name, {[this.#name]: eventName}]
+             of Object.entries(klass[Symbol.metadata].instance.methods)) {
+          if (eventName !== undefined) {
+            data.push({name, eventName});
+          }
+        }
+        klass = klass.__proto__;
+      }
+      this.#map.set(newTarget, data)
+    }
+    return data;
+  }
+}
+
+let handlersMap = new MetadataLookupCache(handler);
+
+function WithActions(superclass) {
+  return class C extends superclass {
+    constructor(...args) {
+      super(...args);
+      let handlers = handlersMap.get(new.target, C);
+      for (const {name, eventName} of handlers) {
+        this.addEventListener(eventName, this[name].bind(this));
+      }
+    }
+  }
 }
-C.prototype.method = f(C.prototype.method);
 ```
 
-`@wrap` can also be used on a class to wrap the entire class.
+#### `@bound` with a mixin
+
+`@bound` could be used with a mixin superclass as follows:
 
 ```js
-@wrap(f)
-class C { }
+class C extends WithBoundMethod(Object) {
+  #x = 1;
+  @bound method() { return this.#x; }
+}
+
+let c = new C;
+let m = c.method;
+m();  // 1, not TypeError
 ```
 
-is roughly equivalent to:
+This decorator could be defined as:
 
 ```js
-class C { }
-C = f(C);
+const boundName = Symbol("boundName");
+function bound(method, context) {
+  context.metadata = {[boundName]: true};
+  return method;
+}
+let boundMap = new MetadataLookupCache(boundName);
+
+function WithBoundMethods(superclass) {
+  return class C extends superclass {
+    constructor(...args) {
+      super(...args);
+      let names = boundMap.get(new.target, C);
+      for (const {name} of names) {
+        this[name] = this[name].bind(this);
+      }
+    }
+  }
+}
 ```
 
-Details:
-- `@wrap` may be used on private methods as well as public ones, static as well as instance.
-- The function is only passed the method, and no other context.
-- The return value is used to replace the method or accessor.
-- `@wrap` may be used on getters or setters, and applies to these individually.
-- `@wrap` may not be used on field declarations, as there's no clear meaning.
-- When `@wrap` is used on a class, if there is a use of `C` in a method or field initializer inside the class, it will refer to the original, unwrapped `C. See [#211](https://github.com/tc39/proposal-decorators/issues/211) for details.
+Note the common use of `MetadataLookupCache` across both examples; this proposal or a follow-on one should consider adding a standard library for adding metadata for this purpose.
 
-### `@register`
+### Option B: The `init` contextual keyword for methods
 
-The `@register` decorator schedules a callback to run after the class is created.
+If it's not acceptable to require a superclass/mixin for cases requiring initialization action, an The `init` keyword in a method declaration changes a method into an "init method". This keyword allows decorators to add initialization actions, run when the constructor executes.
+
+#### `@on` with `init`
```

Comment:
I like this idea! So simple and so powerful, I'm impressed. Though, is there any variants of using something else than `init` keyword? For me, it looks a bit weird when something about decorator is going after the decorator. 

Here are some variants for replace I've come to.

### `@#<decorator name>` decorator
```js
class MyElement extends HTMLElement {
  @#on('click') clickHandler() { }
}
```

I know that many people don't like the `#` symbol for some reason but for me it looks more convenient.

### `@init` decorator

```js
class MyElement extends HTMLElement {
  @init(on, 'click') clickHandler() { }
}
```
Here the `init` is a reserved name for decorator with the following signature:
```ts
function init(fn: Decorator, ...args: Parameters<Decorator>);
```
In case of name conflict the built-in decorator should be used.

Reply:
Yeah, I think you're right, that it would be good to iterate on the syntax here. @rbuckton suggested `@init: on('click')`, which I like personally. What do you think of that syntax?

Reply:
I like that idea better than "a keyword after". It's similarly extensible in the future (`@set: blah`, `@nonenumerable: blah` etc) and feels more attached to the decorator.

Reply:
@littledan, I agree with @ljharb, that's looks promising and reminds me [labeled statements](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/label). 

I also see it as a nice idea for specifying a type of transformation that can be applied. E.g., when we use a simple decorator on a property it is transformed to an accessor pair. But `@set: dec()` can tell the engine to drop the getter and throw an error if the decorator function returns getter. That could bring some parts of the static decorator proposal back and increase the flexibility of decorator system in common.

Reply:
The only thing left is how to combine that decorator labels 😄 (and if it is necessary at all). For me it may look like:
```js
class Foo extends HTMLElement {
  @[nonenumerable, init]: on('click') bar = 1;
}
```

Reply:
@Lodin Might be possible to forego the brackets entirely: `@nonenumerable,init:on("click") bar = 1` isn't ambiguous.

@littledan I could also see that being useful even without other decorators:

```js
@nonenumerable: foo = 1 // `foo` is a non-enumerable instance property
@set: foo = 1 // `foo` is added via `[[Set]]`
```

(My personal style would be eliding whitespace between the comma and `:` in the initial bit, as I view it as part of the process of decoration.)

Reply:
> Might be possible to forego the brackets entirely: `@nonenumerable,init:on("click") bar = 1` isn't ambiguous.

@isiahmeadows, oh I'm not sure I like how it looks. I would prefer some kind of visual container. I thought about an array because we already have annotations in this proposal that look like a JS object, so why not to use an array here. Also, it would be easier for tools like Prettier to make the expression multiline:

```js
class Foo extends HTMLElement {
  @[
    nonenumerable,
    set,
    init,
  ]: on('click')
}
```
vs
```js
class Foo extends HTMLElement {
  @nonenumerable,
    set,
    init: on('click')
}
```
In the second variant, elements do not look like they are part of something whole. I'm not even sure I split it up correctly.

Reply:
I think the decorator name should be in the first place, and all modifications (eg. `init`) should be later because it is more readable. For example, it can be written like this:
```js
class Foo extends HTMLElement {
  @on|nonenumerable|init('click') bar = 1;
}
```

Reply:
@orlov-vo, for me using `:` is better than `|`, and as I mentioned above, for multiple elements we should have the visual container like `[]` because it allows to visually emphasize the enumeration of labels. So it should look like the following:
```js
class Foo extends HTMLElement {
  @on:[nonenumerable,init]('click') bar = 1;
}
```

Basically, I don't have specific objections against that variant of syntax, just thinking that the @rbuckton's approach looks a bit better for me (as I said, it reminds me the [labeled statements](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/label) that are already exist in the language).

Reply:
To be honest, I can't really make sense of the nonenumerable case. It seems like that doesn't expect a decorator after it (it makes perfect sense with no decorator). In general, I'm not really convinced that you want to be able to compose these things, and I don't really know what it would mean in general.